### PR TITLE
Propose only when iteration is less than last block in alt blocks when no. of proposed blocks == max alt blocks

### LINKED ISSUE #95: Dont propose if your iteration < last block in alt blocks if num proposed blocks = max alt blocks
# Rationale
Dont know if this issue exists please verify

we have maxaltblocks parameter in contracts which is currently set at 5

lets say 5 blocks have already been proposed with following iterations
[1,12,100,1000,1232]

now current stakers iteration is 2000. The is no point proposing since the block will not be included in proposed blocks anyway since contract only stores `maxaltblocks` numebr of blocks.

in this case, dont propose to save on gas fees.

### PR DESCRIPTION
**Pull request by Yashk767**
fix #95

### CODE REVIEW

File: cmd/propose.go [hunk 1]
```diff
@@ -59,7 +59,30 @@ func Propose(client *ethclient.Client, account types.Account, config types.Confi
...
+	numOfProposedBlocks, err := utils.GetNumberOfProposedBlocks(client, account.Address, epoch)
+	if err != nil {
+		log.Error(err)
+	}
+	maxAltBlocks, err := utils.GetMaxAltBlocks(client, account.Address)
+	if err != nil {
+		log.Error(err)
+	}
+	if numOfProposedBlocks.Cmp(maxAltBlocks) == 0 || numOfProposedBlocks.Cmp(maxAltBlocks) == 1 {
+		log.Infof("Number of blocks proposed: %s, which is equal or greater than maximum alternative blocks allowed", numOfProposedBlocks)
```
Comment (SkandaBhat) [inline cmd/propose.go:71]: What is Infof?

Comment (ashish10677) [reply cmd/propose.go:71]: `Infof` is like `printf`, a statement that supports string formatting.

File: cmd/propose.go [hunk 2]
```diff
@@ -59,7 +59,30 @@ func Propose(client *ethclient.Client, account types.Account, config types.Confi
...
-
+	numOfProposedBlocks, err := utils.GetNumberOfProposedBlocks(client, account.Address, epoch)
+	if err != nil {
+		log.Error(err)
+	}
+	maxAltBlocks, err := utils.GetMaxAltBlocks(client, account.Address)
+	if err != nil {
+		log.Error(err)
+	}
+	if numOfProposedBlocks.Cmp(maxAltBlocks) == 0 || numOfProposedBlocks.Cmp(maxAltBlocks) == 1 {
```
Comment (ashish10677) [inline cmd/propose.go:70]: Can't you check for >= 0 directly? Instead of checking 2 conditions?

File: cmd/propose.go [hunk 3]
```diff
@@ -59,7 +59,30 @@ func Propose(client *ethclient.Client, account types.Account, config types.Confi
...
+		log.Infof("Number of blocks proposed: %s, which is equal or greater than maximum alternative blocks allowed", numOfProposedBlocks)
+		log.Info("Comparing  iterations...")
+		lastBlockIndex := big.NewInt(0)
+		lastBlockIndex.Sub(numOfProposedBlocks, big.NewInt(1))
```
Comment (ashish10677) [inline cmd/propose.go:74]: This and the above statement can be consolidated into one:
```
lastBlockIndex := big.NewInt(0).Sub(numOfProposedBlocks, big.NewInt(1))
```

File: cmd/propose.go [hunk 4]
```diff
@@ -59,7 +59,30 @@ func Propose(client *ethclient.Client, account types.Account, config types.Confi
...
+		lastProposedBlockStruct, err := utils.GetProposedBlock(client, account.Address, epoch, lastBlockIndex)
+		if err != nil {
+			log.Error(err)
+		}
+		lastIteration := lastProposedBlockStruct.Block.Iteration
+		if lastIteration.Cmp(big.NewInt(int64(iteration))) == -1 {
```
Comment (ashish10677) [inline cmd/propose.go:80]: This is not wrong, but I feel, writing `<0` makes it easier to understand.

**Review by ashish10677 — CHANGES_REQUESTED** [review]

File: cmd/propose.go (hunk 2 above)
Comment (Yashk767) [reply cmd/propose.go:70]: yes, I can do that.

**Review by ashish10677 — APPROVED** [review]

**Pull request merged**